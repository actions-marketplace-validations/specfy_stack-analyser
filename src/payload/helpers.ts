import { listIndexed } from '../common/techs';
import type { AllowedKeys } from '../types/techs';

import { Payload } from '.';

/**
 * When receive a tech in a component, we can deduct a new Component that was missing
 * For example we receive:
 *  - "pg" from a package.json
 *  - we deduct there is a Postgresql
 *  - Create an eponymous component
 *
 * Obviously there could be some false positive.
 */
export function findImplicitComponent(pl: Payload, tech: AllowedKeys) {
  const ref = listIndexed[tech];
  if (ref.type === 'ci' || ref.type === 'language' || ref.type === 'tool') {
    return;
  }

  const comp = new Payload({
    name: tech,
    tech: tech,
    folderPath: pl.path[0],
    parent: pl,
  });
  pl.addComponent(comp);
  if (comp.group !== 'hosting') {
    pl.addEdges(comp.id);
  }
}

/**
 * When receive a tech in a component, we can deduct a new Hosting that was missing
 * For example we receive:
 *  - "vercel" from a folder named .vercel
 *  - we deduct this component is hosted by Vercel
 *  - Create an eponymous component
 *
 * Obviously there could be some false positive.
 */
export function findHosting(pl: Payload, tech: AllowedKeys) {
  const ref = listIndexed[tech];
  if (ref.type !== 'hosting') {
    return;
  }

  const find = pl.components.find((c) => c.tech === ref.key);
  if (!find) {
    throw new Error(`cant find hosting ${ref.key}`);
  }

  pl.inComponent = find.id;
}

export function findEdges(pl: Payload) {
  pl.components.forEach((component) => {
    component.techs.forEach((tech) => {
      const ref = listIndexed[tech];

      if (
        ref.type === 'sass' ||
        ref.type === 'app' ||
        ref.type === 'db' ||
        ref.type === 'messaging'
      ) {
        const find = pl.components.find((c) => c.tech === ref.key);
        if (!find) {
          throw new Error(`cant find sass ${ref.key}`);
        }
        component.addEdges(find.id);

        return;
      }
    });
  });
}
