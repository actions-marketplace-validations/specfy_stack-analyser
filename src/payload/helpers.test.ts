import { describe, expect, it } from 'vitest';

import { findEdgesInDependencies, flatten } from './helpers';

import { Payload } from '.';

describe('findEdgesInDependencies', () => {
  it('should find an edge', () => {
    const main = new Payload({ id: '1', name: 'main', folderPath: '' });
    main.childs.push(new Payload({ id: '2', name: '2', folderPath: '/' }));
    main.childs.push(
      new Payload({
        id: '3',
        name: '3',
        folderPath: '/',
        dependencies: [['npm', '2', '0.0.0']],
      })
    );

    findEdgesInDependencies(main);

    expect(main).toMatchSnapshot();
  });
});

describe('flatten', () => {
  it('should flatten', () => {
    const main = new Payload({ id: '1', name: 'main', folderPath: '' });
    const sub = new Payload({
      id: '2',
      name: '2',
      folderPath: '/',
      dependencies: [['npm', 'aa', '0.0.0']],
    });
    sub.childs.push(
      new Payload({
        id: '3',
        name: '3',
        folderPath: '/',
        dependencies: [['npm', 'bb', '0.0.0']],
      })
    );
    main.childs.push(sub);

    const flat = flatten(main);
    expect(flat.id).not.toBe(main.id);
    flat.id = 'new';

    expect(flat.toJson()).toMatchSnapshot();
  });
});
