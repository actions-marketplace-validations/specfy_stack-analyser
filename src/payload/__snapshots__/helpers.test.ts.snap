// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findEdgesInDependencies > should find an edge 1`] = `
Payload {
  "childs": [
    Payload {
      "childs": [],
      "dependencies": [],
      "edges": [],
      "group": "component",
      "id": "2",
      "inComponent": null,
      "languages": {},
      "name": "2",
      "parent": undefined,
      "path": [
        "/",
      ],
      "tech": null,
      "techs": Set {},
    },
    Payload {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "2",
          "0.0.0",
        ],
      ],
      "edges": [
        {
          "portSource": "right",
          "portTarget": "left",
          "read": true,
          "to": Payload {
            "childs": [],
            "dependencies": [],
            "edges": [],
            "group": "component",
            "id": "2",
            "inComponent": null,
            "languages": {},
            "name": "2",
            "parent": undefined,
            "path": [
              "/",
            ],
            "tech": null,
            "techs": Set {},
          },
          "vertices": [],
          "write": true,
        },
      ],
      "group": "component",
      "id": "3",
      "inComponent": null,
      "languages": {},
      "name": "3",
      "parent": undefined,
      "path": [
        "/",
      ],
      "tech": null,
      "techs": Set {},
    },
  ],
  "dependencies": [],
  "edges": [],
  "group": "component",
  "id": "1",
  "inComponent": null,
  "languages": {},
  "name": "main",
  "parent": undefined,
  "path": [
    "",
  ],
  "tech": null,
  "techs": Set {},
}
`;

exports[`flatten > should flatten 1`] = `
{
  "childs": [
    {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "aa",
          "0.0.0",
        ],
      ],
      "edges": [],
      "group": "component",
      "id": "2",
      "inComponent": null,
      "languages": {},
      "name": "2",
      "path": [
        "/",
      ],
      "tech": null,
      "techs": [],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "bb",
          "0.0.0",
        ],
      ],
      "edges": [],
      "group": "component",
      "id": "3",
      "inComponent": null,
      "languages": {},
      "name": "3",
      "path": [
        "/",
      ],
      "tech": null,
      "techs": [],
    },
  ],
  "dependencies": [
    [
      "npm",
      "bb",
      "0.0.0",
    ],
    [
      "npm",
      "aa",
      "0.0.0",
    ],
  ],
  "edges": [],
  "group": "component",
  "id": "new",
  "inComponent": null,
  "languages": {},
  "name": "flatten",
  "path": [
    "/",
    "",
  ],
  "tech": null,
  "techs": [],
}
`;
